File: ruby/tournament/tournament.rb
class Tournament

  attr_reader :matches
  def self.tally(input)
    @teams = []
    @matches = []

    score_list = ["Team                           | MP |  W |  D |  L |  P\n"]

    input.split("\n").map { |e| e.split(';') }.each do |m|
      team_1 = create_or_find_team(m[0])
      team_2 = create_or_find_team(m[1])
      @matches << Match.new(team_1, team_2, m[2])
    end

    @teams.sort { |a, b| [b.points, a.name] <=> [a.points, b.name] }.each do |t|
      score_list << "%-31s|%3d |%3d |%3d |%3d |%3d\n" % [t.name, t.matches, t.wins, t.draws, t.loses, t.points]
    end

    score_list.join
  end


  def self.create_or_find_team(team_name)
    team = @teams.find { |e| e.name == team_name }
    return team unless team.nil?
    team = Team.new(team_name)
    @teams << team
    team
  end


  class Team
    attr_accessor :wins, :draws, :loses
    attr_reader :name
    def initialize(name)
      @name = name
      @wins, @draws, @loses= 0, 0, 0
    end

    def points
      @wins * 3 + @draws * 1
    end

    def matches
      @wins + @loses + @draws
    end
  end

  class Match
    attr_reader :team_1, :team_2, :result
    def initialize(team_1, team_2, result)
      @team_1 = team_1
      @team_2 = team_2
      @result = result
      case result
      when 'win'
        @team_1.wins += 1
        @team_2.loses += 1
      when 'draw'
        @team_1.draws += 1
        @team_2.draws += 1
      when 'loss'
        @team_1.loses += 1
        @team_2.wins += 1
      else
        # invalid result do nothing
      end
    end
  end

end

module BookKeeping
  VERSION = 3
end

File: python/tournament/tournament.py
import re

def tally(rows):
    header = 'Team                           | MP |  W |  D |  L |  P'
    ranks = [header]
    teams = {}
    for row in rows:
        teams = add_match_result(teams, row)
    
    ranks += calculate_ranks(teams)
    return ranks

def calculate_ranks(teams):
    formatter = '{:<31s}|{:>3d} |{:>3d} |{:>3d} |{:>3d} |{:3d}'
    ranks = []
    for t in sorted(teams.items(), key=lambda v: (-v[1]['P'], v[0])):
        ranks.append(formatter.format(t[0], t[1]['MP'], t[1]['W'], t[1]['D'], t[1]['L'], t[1]['P']))
    return ranks

def add_match_result(teams, row):
    m = re.match('(.*?);(.*?);(.*)$', row)
    home_team_result = 'win' if m.group(3) == 'loss' else ('loss' if m.group(3) == 'win' else m.group(3))
    teams = add_team_result(m.group(1), m.group(3), teams)
    teams = add_team_result(m.group(2), home_team_result, teams)
    return teams

def add_team_result(team, result, teams):
    if team in teams.keys():
        teams[team]['MP'] += 1
        teams[team]['W'] += 1 if result == 'win' else 0
        teams[team]['D'] += 1 if result == 'draw' else 0
        teams[team]['L'] += 1 if result == 'loss' else 0
        teams[team]['P'] += 3 if result == 'win' else (1 if result == 'draw' else 0)
    else:
        teams[team] = {'MP': 1}
        teams[team]['W'] = 1 if result == 'win' else 0
        teams[team]['D'] = 1 if result == 'draw' else 0
        teams[team]['L'] = 1 if result == 'loss' else 0
        teams[team]['P'] = 3 if result == 'win' else (1 if result == 'draw' else 0)
    return teams
